added interrupt based bus-off recovery

raffael0 opened the pull request:
This PR allows the ECUs to recover their CAN communication ability after they are sent into the bus-off state due to too many errors.  

Should be merged together with [firmware_liquids](https://github.com/SpaceTeam/firmware_liquids/pull/5)

raffael0 committed:
Added interrupt based bus-off recovery. An Interrupt is now triggered
when the Bus-Off state is changed and the device reintegrates into the
Bus

miDeb commented on Src/STRHAL_CAN.c:
```diff
@@ -470,6 +482,11 @@ void FDCAN1_IT0_IRQHandler(void)
 		FDCAN1->RXF1A = i & 0x7;
 	}
+	if (FDCAN1->IR & FDCAN_IR_BO) {
+		SET_BIT(FDCAN1->IR, FDCAN_IR_BO);
```
Maybe add a comment that writing 1 to this bit actually clears it.

miDeb commented on Src/STRHAL_CAN.c:
```diff
@@ -470,6 +482,11 @@ void FDCAN1_IT0_IRQHandler(void)
 			rec(id, rx_fifo->data.byte, l - 2);
 		FDCAN1->RXF1A = i & 0x7;
 	}
+	if (FDCAN1->IR & FDCAN_IR_BO) {
```
It looks like this is triggered every time the bus-off state changes, not just when entering it:
```
Bit 19 BO: Bus-off status
0: Bus-off status unchanged
1: Bus-off status changed
```

But I don't think this causes a problem, right? Also reading `PSR` to check if we are currently in bus off state wouldn't be ideal because we would reset some of the "clear on read" bits.

raffael0 replied:
Good catch. It worked when I tested it but I can also add another check to see if the CCR Init Bit is set. That's how I did it in the Ecu_Lamarr.cpp implementation

miDeb reviewed (approved):
Looks good

raffael0 committed:
Fixed Bug in FDCAN2 Message handling where the interrupt wasn't properly cleared

raffael0 commented:
I added a check to see if we are in the bus-off state and also fixed a bug in the FDCAN2 Interrupt handler where the interrupts aren't properly cleared. I tested it on the hardware so I'm ready to merge on my end (I want to merge without squashing to highlight the bugfix).

miDeb reviewed (approved):
Nice catch, sounds good to me
